Updated Backup Policy documentation

PacoDw opened the pull request:
## Description

Updated backup_policy documentation explaining how we can implement at least one policy item or remove ones.

Link to any related issue(s): #222 

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the MongoDB CLA
- [x] I have read the Terraform contribution guidelines 
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirments
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments

PacoDw committed:
doc: updated backup_policy resource explaining on how I can implement at least one policy item

themantissa commented on website/docs/r/cloud_provider_snapshot_backup_policy.html.markdown:
````diff
@@ -73,6 +73,47 @@ resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {
 }
 ```
 
+## Example Usage implementing at least one Policy Item
````
So this is taking me a while to review and after testing it all out I think we need 3 examples:

## Example Usage - Create a Cluster and Modify the 4 Default Policies Simultaneously
_This is the existing example we already had but just need to change the header_

## Example Usage - Create a Cluster and Modify 3 Default Policies and Remove 1 Default Policy Simultaneously
_This needs to be a new example, similar to the existing example that just removes one of the default policy blocks_

## Example Usage - Remove 3 Default Policies Items After the Cluster Has Already Been Created
_Same example that's already here._  
 
Each example also needs a note to explain what it is doing:
First example:
-> **NOTE:** When Cloud Backup is enabled for a cluster MongoDB Atlas automatically creates a default Cloud Backup schedule for the cluster with four policy items; hourly, daily, weekly, and monthly.   Because of this default creation this provider automatically saves the Cloud Backup Snapshot Policy into the Terraform state.  If the default works well for you then you do not need to do anything other than create a cluster with Cloud Backup enabled and your Terraform state will have this information if you need it. However,  if you want the policy to be different than the default when creating the cluster simply follow this example to modify the four default policies.

Second example:
-> **NOTE:** (See text in first example for more details on the default.) If you want the Cloud Backup Snapshot Policy to vary in the number of policies from the default when creating the cluster, perhaps you want to remove one policy item and modify the remaining three, simply follow this example here to remove a policy and modify three.

Third example:
-> **NOTE:** (See text in first example for more details on the default.) If you want the Cloud Backup Snapshot Policy to vary in number of policies for a cluster that was already created/imported, perhaps you want to remove three policy items and modify the remaining policy, simply follow the example here.

PacoDw replied:
I like it it's a great idea. I will implement it, thanks. 👍

themantissa commented on website/docs/r/cloud_provider_snapshot_backup_policy.html.markdown:
````diff
@@ -73,6 +73,47 @@ resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {
+## Example Usage implementing at least one Policy Item
+
+```hcl
+resource "mongodbatlas_cluster" "my_cluster" {
````
I think a comment here explaining "this cluster has already been created and is here as an example" would help.

themantissa commented on website/docs/r/cloud_provider_snapshot_backup_policy.html.markdown:
```diff
@@ -73,6 +73,47 @@ resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {
+    id = mongodbatlas_cluster.my_cluster.snapshot_backup_policy.0.policies.0.id
+
+    policy_item {
+      id                 = 5f0747cad187d8609a72f546
```
This needs some explanation since it's a hard coded value.   Perhaps this would do it:

This is the id MongoDB Atlas returns for the policy item we want to keep.  Here it is hard coded because you need to either use the actual value from the Terraform state or look to map the policy item you want to keep to it's placement in the state file array that was imported in when the cluster was originally created.  For example in the state file we are using the hourly policy is the first policy in the array so it can be referred to with
mongodbatlas_cluster.my_cluster.snapshot_backup_policy.0.policies.0.policy_item.0.id instead of the hard coded value.  Summarized: to use the state file value for a policy item you need to determine the array placement # of the same frequency_type you want to keep.

themantissa replied:
This was resolved but it's still a hard coded value that would be hard for someone to understand where it comes from.

themantissa reviewed (changes requested):
This helped me a great deal, as did running through the examples!   I've included quite a bit here that I think will further clear this up.  Let me know what you think.

PacoDw committed:
chore: added more examples about on how modify the policy items

themantissa commented on website/docs/r/cloud_provider_snapshot_backup_policy.html.markdown:
```diff
@@ -72,6 +75,197 @@ resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {
+When Cloud Backup is enabled for a cluster MongoDB Atlas automatically creates a default Cloud Backup schedule for the cluster with four policy items; hourly, daily, weekly, and monthly. Because of this default creation this provider automatically saves the Cloud Backup Snapshot Policy into the Terraform state. If the default works well for you then you do not need to do anything other than creat ...
+
+## Example Usage - Create a Cluster and Modify the 4 Default Policies Simultaneously
+This cluster has already been created and is here as an example
```
The headings and the examples are a bit mixed up.  Simultaneously means "at the same time" so this heading and example don't match.  The heading as shown here would mean "you create a cluster and change that clusters default backup policy in the same Terraform apply run" - that is the first example on lines 15-77.  So I think the solve here is to use this heading for the first example and then remove this one.

themantissa commented on website/docs/r/cloud_provider_snapshot_backup_policy.html.markdown:
```diff
@@ -72,6 +75,197 @@ resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {
+      retention_value    = 2
+    }
+
+    # Item removed
+    # policy_item {
+    #   id                 = mongodbatlas_cluster.my_cluster.snapshot_backup_policy.0.policies.0.policy_item.2.id
+    #   frequency_interval = 4
+    #   frequency_type     = "weekly"
+    #   retention_unit     = "weeks"
+    #   retention_value    = 3
+    # }
```
This is a very good way to illustrate it!  Nice!

themantissa commented on website/docs/r/cloud_provider_snapshot_backup_policy.html.markdown:
````diff
@@ -72,6 +75,197 @@ resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {
 }
 ```
+# Examples Modifying Polices
+When Cloud Backup is enabled for a cluster MongoDB Atlas automatically creates a default Cloud Backup schedule for the cluster with four policy items; hourly, daily, weekly, and monthly. Because of this default creation this provider automatically saves the Cloud Backup Snapshot Policy into the Terraform state. If the default works well for you then you do not need to do anything other than creat ...
````
This should go to the very top as all of the examples have to do w/ modifying as the initial policy is created by default.

themantissa commented on website/docs/r/cloud_provider_snapshot_backup_policy.html.markdown:
```diff
@@ -72,6 +75,197 @@ resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {
+
+-> **NOTE:** This is the id MongoDB Atlas returns for the policy item we want to keep. Here it is hard coded because you need to either use the actual value from the Terraform state or look to map the policy item you want to keep to it's placement in the state file array that was imported in when the cluster was originally created.
+
+Summarized: to use the state file value for a policy item you need to determine the array placement # of the same frequency_type you want to keep. With this Terraform configuration the anothers policy items will be removed.
```
The NOTE and IMPORTANT both refer to line 256, the hard coded policy_item id - we have to explain to the user where the hard coded value comes from and how to figure this stuff out.  It's all probably best put in a comment in the example above the value.

themantissa commented on website/docs/r/cloud_provider_snapshot_backup_policy.html.markdown:
```diff
@@ -72,6 +75,197 @@ resource "mongodbatlas_cloud_provider_snapshot_backup_policy" "test" {
+
+-> **NOTE:** (See text in first example for more details on the default.) If you want the Cloud Backup Snapshot Policy to vary in number of policies for a cluster that was already created/imported, perhaps you want to remove three policy items and modify the remaining policy, simply follow the above example here.
+
+~> **IMPORTANT:** Note in this example we decided to remove the first 3 items so we can't use `mongodbatlas_cluster.my_cluster.snapshot_backup_policy.0.policies.0.policy_item.3.id` this sentence to retrieve the monthly id value of the cluster state due to once the cluster being modified or makes a `terraform refresh` will cause that the three items will remove from the state, so we will get an er ...
```
this should be under the heading, Example Usage - Remove 3 Default Policies Items After the Cluster Has Already Been Created, as it is explaining what the example is.

The IMPORTANT probably is best to put near 256 in a comment or use the other explanation cut and pasted in here.

themantissa reviewed (changes requested):
The examples are good but some of the placement of the provided info is not in the best place.  Let me know if this makes sense and I can perhaps submit a PR with the changes if that will help. 
Just FYI - I don't want this to hold up 0.6.2 but also want to get better examples out - we may need to do 0.6.2 first and then do a special push for just this after.

PacoDw committed:
chore: changed by requested changes

themantissa committed:
Update cloud_provider_snapshot_backup_policy.html.markdown

Modified PR branch to quicken pace of completion so we can proceed with repo move.

themantissa commented:
@PacoDw this is a little non-standard but I made the changes I think we need in the last commit.  Let me know if they make sense and then I can approve so we can move forward on closing these out.  Thanks!

PacoDw commented:
@themantissa no problem it's okay 👍

themantissa reviewed (approved):
LGTM (since made updates in last commit I think we still need).  Let's wrap this up.

coderGo93 reviewed (approved):
LGTM
